### DISCUSSION TITLE
How does jQuery detect the vendor prefixes to add?

### DISCUSSION THREAD

**Discussion by vlakoff** [jquery/jquery#5516] [Q&A] [answered]
I have some code that does `.css( 'user-select', 'none' )`.

A great thing is that jQuery automatically converts the property name to `-moz-user-select` when using [Pale Moon](https://en.wikipedia.org/wiki/Pale_Moon) browser, and (presumably) to `-webkit-user-select` when using [Safari](https://en.wikipedia.org/wiki/Safari_(web_browser)) browser.

I had a look at jQuery source code, but did not figure out how jQuery manages to intelligently add vendor prefixes when needed. I'm pretty sure there is no hardcoded list of properties/browsers, so some detection should be made somewhere… Could anyone tell how the magic happens?

(By the way, Safari is the only remaining major browser that doesn't support the unprefixed [`user-select`](https://caniuse.com/?search=user-select) CSS property. I have already sent them feedback messages about this oversight, but if you have better channels to reach them, you may want to give it a try.)

**Comment by mgol** [comment, marked as answer]
The logic is contained in https://github.com/jquery/jquery/blob/1fa8df5dbd5d84cf55882a38eb6e571abd0aa938/src/css/finalPropName.js, in the `finalPropName` function. It keeps an object caching mapping from a property to its prefixed versions and consults it at the beginning:
https://github.com/jquery/jquery/blob/1fa8df5dbd5d84cf55882a38eb6e571abd0aa938/src/css/finalPropName.js#L24-L28
If the cache does not exist, it checks if the property itself is supported by checking if it's a key in an empty style object:
https://github.com/jquery/jquery/blob/1fa8df5dbd5d84cf55882a38eb6e571abd0aa938/src/css/finalPropName.js#L29-L31
If not, it tries to guess a vendor prefix by trying prefixed versions of the supplied property:
https://github.com/jquery/jquery/blob/1fa8df5dbd5d84cf55882a38eb6e571abd0aa938/src/css/finalPropName.js#L8-L20
The list of accepted vendor prefixes is hardcoded:
https://github.com/jquery/jquery/blob/1fa8df5dbd5d84cf55882a38eb6e571abd0aa938/src/css/finalPropName.js#L3

It's important that this logic is applied for every property in isolation as some browsers now support multiple vendor prefixes; if we cached the prefix globally, we could be checking a wrong one for a different property.

**Comment by vlakoff** [comment]
Coming back a bit late. Thank you for the detailed explanation. Now that the relevant file is found, things become much simpler to understand.

Side note (no action needed):

I noticed that, due to the [`in` operators](https://developer.mozilla.org/en-US/docs/Web/JavaScript/Reference/Operators/in) (lines 16 and 29), the code would be erroneous if looking for CSS property names that are also property names of the [`style`](https://developer.mozilla.org/en-US/docs/Web/API/HTMLElement/style) (instance of [`CSSStyleDeclaration`](https://developer.mozilla.org/en-US/docs/Web/API/CSSStyleDeclaration)) JavaScript object.

But that's only a theoretical issue. JavaScript has historically been preventing collisions between HTML/CSS names and JavaScript object property names (the most notable example being `class` HTML attribute → `className` in JavaScript). Nevertheless, if ever the user queried such a name, the code would simply make a "false positive" and return the input value as-is, unprefixed.

**Comment by vlakoff** [comment]
However, I noticed that when querying a CSS property name that exists unprefixed, it doesn't get cached into the `vendorProps` map (contrarily to the prefixed names, that get cached).

Therefore, when querying unprefixed names, each time there is one lookup in the `vendorProps` map then one lookup in the `emptyStyle` object, whereas we could save the latter (after the cache is populated on first access).

Namely, I'm suggesting this change:
```js
export function finalPropName( name ) {
	var final = vendorProps[ name ];

	if ( final ) {
		return final;
	}
	if ( name in emptyStyle ) {
		// replace:
		//return name;
		// with:
		return vendorProps[ name ] = name;
	}
	return vendorProps[ name ] = vendorPropName( name ) || name;
}
```

Then to avoid confusion, we should rename the `vendorProps` map to something like `finalProps`.

(Just as a reminder: The user might also query already prefixed names. Such use case would, and should, work as well. Just something to keep in mind.)

**Comment by vlakoff** [comment]
I opened an issue about this: #5582.
